src/units/abilities.cpp

```diff
@@ -1146,14 +1168,168 @@ unit_ability_list attack_type::get_special_ability(const std::string& ability) c
 	return abil_list;
 }
 
-bool attack_type::bool_ability(const std::string& ability) const
+static bool get_ability_children(std::vector<special_match>& tag_result,
+	                           std::vector<special_match>& id_result,
+	                           const config& parent, const std::string& id,
+	                           bool special_id=true, bool special_tags=true) {
+		if(special_id && special_tags){
+			if ( get_special_children(tag_result, id_result, parent, id) ) {
+				return true;
+			}
+		} else if(special_id && !special_tags){
+			if ( get_special_children_id(id_result, parent, id) ) {
+				return true;
+			}
+		} else if(!special_id && special_tags){
+			if ( get_special_children_tags(tag_result, parent, id) ) {
+				return true;
+			}
+		}
+		return false;
+}
+
+bool attack_type::get_special_ability_bool(const std::string& special, bool special_id, bool special_tags) const
 {
-	bool abil_bool = get_special_bool(ability);
-	unit_ability_list abil = list_ability(ability);
-	if(!abil.empty()) {
-		abil_bool = true;
+	const unit_map& units = display::get_singleton()->get_units();
+
+	unit_const_ptr self = self_;
+	unit_const_ptr other = other_;
+
+	if(self == nullptr) {
+		unit_map::const_iterator it = units.find(self_loc_);
+		if(it.valid()) {
+			self = it.get_shared_ptr();
+		}
 	}
-	return abil_bool;
+	if(other == nullptr) {
+		unit_map::const_iterator it = units.find(other_loc_);
+		if(it.valid()) {
+			other = it.get_shared_ptr();
+		}
+	}
+
+	// Make sure they're facing each other.
+	temporary_facing self_facing(self, self_loc_.get_relative_dir(other_loc_));
+	temporary_facing other_facing(other, other_loc_.get_relative_dir(self_loc_));
+	const map_location loc = self_ ? self_->get_location() : self_loc_;
+	unit_ability_list abil_list(loc);
+	unit_ability_list abil_list_id(loc);
+	unit_ability_list abil_other_list(loc);
+	unit_ability_list abil_other_list_id(loc);
+	assert(display::get_singleton());
+	if(self){
+		std::vector<special_match> special_tag_matches;
+		std::vector<special_match> special_id_matches;
+		if(get_ability_children(special_tag_matches, special_id_matches, (*self).abilities(), special, special_id , special_tags)){
+			return true;
+		}
+		
+		adjacent_loc_array_t adjacent;
+		get_adjacent_tiles(self_loc_,adjacent.data());
+		for(unsigned i = 0; i < adjacent.size(); ++i) {
+			const unit_map::const_iterator it = units.find(adjacent[i]);
+			if (it == units.end() || it->incapacitated())
+				continue;
+			if ( &*it == self.get() )
+				continue;
+
+			if(get_ability_children(special_tag_matches, special_id_matches, it->abilities(), special, special_id , special_tags)){
+				return true;
+			}
+		}
+		if(special_tags){
+			for(const special_match& entry : special_tag_matches) {
+				if(entry.tag_name !="heals" && entry.tag_name !="regenerate" && entry.tag_name !="skirmisher" && entry.tag_name !="teleport" && entry.tag_name !="hides"){
```

Comment:
I think it would be nicer to put this list of excluded tags into a `std::set` and then just call `excluded_tags.count(entry.tag_name) == 0`.

Reply:
it is done